crates/ide/src/link_rewrite.rs

```diff
@@ -79,3 +134,109 @@ fn map_links<'e>(
         _ => evt,
     })
 }
+
+fn parse_link(s: &str) -> (&str, Option<hir::Namespace>) {
+    let path = strip_prefixes_suffixes(s);
+    let ns = ns_from_intra_spec(s);
+    (path, ns)
+}
+
+/// Strip prefixes, suffixes, and inline code marks from the given string.
+fn strip_prefixes_suffixes(mut s: &str) -> &str {
+    s = s.trim_matches('`');
+
+    [
+        (TYPES.0.iter(), TYPES.1.iter()),
+        (VALUES.0.iter(), VALUES.1.iter()),
+        (MACROS.0.iter(), MACROS.1.iter()),
+    ]
+    .iter()
+    .for_each(|(prefixes, suffixes)| {
+        prefixes.clone().for_each(|prefix| s = s.trim_start_matches(*prefix));
+        suffixes.clone().for_each(|suffix| s = s.trim_end_matches(*suffix));
+    });
+    s.trim_start_matches("@").trim()
+}
+
+static TYPES: ([&str; 7], [&str; 0]) =
+    (["type", "struct", "enum", "mod", "trait", "union", "module"], []);
+static VALUES: ([&str; 8], [&str; 1]) =
+    (["value", "function", "fn", "method", "const", "static", "mod", "module"], ["()"]);
+static MACROS: ([&str; 1], [&str; 1]) = (["macro"], ["!"]);
+
+/// Extract the specified namespace from an intra-doc-link if one exists.
+///
+/// # Examples
+///
+/// * `struct MyStruct` -> `Namespace::Types`
+/// * `panic!` -> `Namespace::Macros`
+/// * `fn@from_intra_spec` -> `Namespace::Values`
+fn ns_from_intra_spec(s: &str) -> Option<hir::Namespace> {
+    [
+        (hir::Namespace::Types, (TYPES.0.iter(), TYPES.1.iter())),
+        (hir::Namespace::Values, (VALUES.0.iter(), VALUES.1.iter())),
+        (hir::Namespace::Macros, (MACROS.0.iter(), MACROS.1.iter())),
+    ]
+    .iter()
+    .filter(|(_ns, (prefixes, suffixes))| {
+        prefixes
+            .clone()
+            .map(|prefix| {
+                s.starts_with(*prefix)
+                    && s.chars()
+                        .nth(prefix.len() + 1)
+                        .map(|c| c == '@' || c == ' ')
+                        .unwrap_or(false)
+            })
+            .any(|cond| cond)
+            || suffixes
+                .clone()
+                .map(|suffix| {
+                    s.starts_with(*suffix)
+                        && s.chars()
+                            .nth(suffix.len() + 1)
+                            .map(|c| c == '@' || c == ' ')
+                            .unwrap_or(false)
+                })
+                .any(|cond| cond)
+    })
+    .map(|(ns, (_, _))| *ns)
+    .next()
+}
+
+fn get_doc_url(db: &RootDatabase, krate: &Crate) -> Option<Url> {
+    krate
+        .get_html_root_url(db)
+        .or_else(|| {
+            // Fallback to docs.rs. This uses `display_name` and can never be
+            // correct, but that's what fallbacks are about.
+            //
+            // FIXME: clicking on the link should just open the file in the editor,
+            // instead of falling back to external urls.
+            Some(format!("https://docs.rs/{}/*/", krate.display_name(db)?))
```

Comment:
I disagree that we should fallback to opening the file here. It should definitely be an option in general, but the fallback should try to match the behaviour of the rest of the code.